### PR TITLE
Build config and Debug mode

### PR DESCRIPTION
**Pull request by Jolanrensen**
This PR adds the [buildconfig](https://github.com/gmazzo/gradle-buildconfig-plugin) plugin to the project.

This achieves two things:
- It allows us to access the gradle project version via `BuildConfig.VERSION`. We can use this well in the Jupyter integration where we're often interested which dataframe version is pulled into the notebook when doing a `%use dataframe`. 
- It helps with https://github.com/Kotlin/dataframe/issues/713. 
  By simply adding `-Pkotlin.dataframe.debug=true` to the build on TC, we can run the extra checks from the issue. By default this check should be off, as it costs performance, but it's very valuable for catches bugs. We can use `BuildConfig.DEBUG` in other places in the future as well.

See each commit for more info.

We should add the flag to TC only after https://github.com/Kotlin/dataframe/issues/713 is solved completely, otherwise all builds will fail.

### CODE REVIEW

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt [hunk 1]
```diff
@@ -12,6 +17,32 @@ internal abstract class DataColumnImpl<T>(
...
+                    this!!::class.qualifiedName == type.classifier?.let { (it as KClass<*>).qualifiedName }
+
+            this.isArray -> type.isArray
+
+            // cannot check the precise type of array
+            else -> this!!::class.isSubclassOf(type.classifier as KClass<*>)
+        }
+
+    init {
+        /* Check for [Issue #713](https://github.com/Kotlin/dataframe/issues/713).
```
Comment (zaleslaw) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt:35]: strange formatting of KDoc

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt [hunk 2]
```diff
@@ -12,6 +17,32 @@ internal abstract class DataColumnImpl<T>(
...
+         * This only runs with `kotlin.dataframe.debug=true` in gradle.properties
+         */
+        if (BuildConfig.DEBUG) {
+            require(values.all { it matches type }) {
```
Comment (zaleslaw) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt:39]: It checks all values, right? Is it worth to check random 10-20 numbers from the column?

File: gradle.properties [hunk 3]
```diff
@@ -11,3 +11,8 @@ org.gradle.jvmargs=-Xmx4G
 # This makes it mandatory to explicitly apply your own version of the
 # KSP plugin in the modules that use it.
 kotlin.dataframe.add.ksp=false
+
+# Enables debug mode for dataframe.
+# This can make certain tests run that should not be run in production.
+# It can also be turned on from the command line with `-Pkotlin.dataframe.debug=true`
+kotlin.dataframe.debug=false
```
Comment (zaleslaw) [inline gradle.properties:18]: does it mean, that we could add more modes, with different names, it's not a strongly-defined schema?

**Review by zaleslaw — APPROVED** [review]

File: gradle.properties (hunk 3 above)
Comment (Jolanrensen) [reply gradle.properties:18]: not strong typed indeed, just an argument I came up with. Similar to the add.ksp argument.
It's not strongly typed, but when read it tries to parse the argument as Boolean and if failed it defaults to `false`

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt (hunk 2 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt:39]: If we only check a handful of values there's a chance we won't catch the outlier and get flaky tests.

**Comment by Jolanrensen** [comment]
Thanks!

**Pull request merged**